docs/honk.1
.\"
.\" Copyright (c) 2019 Ted Unangst
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt HONK 1
.Os
.Sh NAME
.Nm honk
.Nd federated status conveyance
.Sh DESCRIPTION
The
.Nm
utility processes federated status updates and other microblog activities.
This is the user manual.
For administration, see
.Xr honk 8 .
For other documentation, refer to the
.Xr intro 1 .
.Pp
This manual is still incomplete.
It'll get there eventually.
.Ss Honkers
Initially, there won't be much to see after logging in.
In order to receive regular updates from other users, they must first
be added to one's honker collection.
Begin at the
.Pa honkers
tab.
The
.Ar url
field is required.
Either of two forms are accepted, the user's handle (or webfinger) or their
ActivityPub actor URL.
.Pp
.Dl @user@example.social
.Dl https://example.social/users/user
.Pp
The
.Ar name
field is optional and will be automatically inferred.
The
.Ar notes
field is reserved for user remarks.
Fellow honkers may be added to one or more
.Ar combos
to suit one's organizational preferences.
These are accessed via the
.Pa combos
tab and allow easy access to particular groupings.
The special combo name of one hyphen
.Sq -
will exclude a honker's posts from the primary feed.
.Pp
It is also possible to skip subscribing.
In this case, regular posts are not received, but replies and posts fetched
via other means will appear in the relevant combos.
.Pp
In addition to honkers, it is possible to subscribe to a hashtag collection.
(Where supported.)
Enter the collection URL for
.Ar url .
.Pp
Separately, hashtags may be added to a combo by creating a honker with a
.Ar url
of the desired hashtag (including #).
Several hashtags may thus be collected in a single combo.
.Ss Viewing
The primary feed is accessed via the
.Pa home
tab.
It will contain posts from all honkers except those specifically excluded.
Posts mentioning the user, both followed and not, are collected under the
.Pa @me
tab.
Other feeds include
.Pa first
which excludes replies, the user defined options under the
.Pa combos
subheading, and the
.Pa events
page which lists only events.
.Pp
Individual honks contain a visual representation of the honker's ID,
their name, the activity (with a link back to origin), a link to the
parent post if applicable, and the convoy (thread) identifier.
A red border indicates the honk is not public.
Screenshot below.
.Pp
.Lk screenshot-honk.png screenshot of one honk
.Pp
Available actions are:
.Bl -tag -width tenletters
.It Ic bonk
Share with followers.
Not available for nonpublic honks.
.It Ic honk back
Reply.
.It Ic mute
Mute this entire thread.
Existing posts are hidden, and future posts will not appear in any feed.
.It Ic zonk
Delete this post.
When deleting one's own post, other servers will be requested to remove it,
but this is unreliable.
.It Ic ack
Acknowledge reading this post.
Typically if it's a reply to one's own post.
.It Ic save
Save this honk to the
.Pa saved
tab to find later.
.It Ic untag me
Sometimes a thread goes on entirely too long.
Untag will hide further replies to the selected post, but without muting the
entire thread.
Replies higher in the tree are still received.
.It Ic badonkadonk
Please no.
.It Ic edit
Change it up.
Alas, Update activities do not federate reliably.
.Ss Refresh
Clicking the refresh button will load new honks, if any.
New honks will be subtly highlighted.
.El
.Ss Honking
Refer to the
.Xr honk 5
section of the manual for details of honk composition.
.Ss Search
Find old honks.
It's basic substring match with a few extensions.
The following keywords are supported:
.Bl -tag -width honker
.It site
Substring match on the post domain name.
.It honker
Exact match, either AP actor or honker nickname.
.It -
Negate term.
.El
.Pp
Example:
.Dl honker:goose big moose -footloose
This query will find honks by the goose about the big moose, but excluding
those about footloose.
.Ss Filtering
Sometimes other users of the federation can get unruly.
The honk filtering and censorship system,
.Xr hfcs 1 ,
can be of great use to restore order to one's timeline.
Accessed via the
.Pa filters
menu item.
.Ss Xzone
The
.Pa xzone
page lists recently seen honkers that are not otherwise tracked.
It also allows the import of external objects via URL, either individual
posts or actor URLs, in which case their recent outbox is imported.
.Ss Account
It's all about you.
An avatar may be selected from the
.Pa funzone
by adding
.Dq avatar: filename.png
to one's profile info.
If truly necessary.
.Pp
Some options to customize the site appearance:
.Bl -tag -width skinny
.It skinny
Use a narrower column for the main display.
.It omit images
Omit img tags, to lighten page loads on slow connections.
.It apple
Prefer Apple links for maps.
The default is OpenStreetMap.
.It reaction
Pick an emoji for reacting to posts.
.El
.Sh ENVIRONMENT
.Nm
is designed to work with most browsers, but for optimal results it is
recommended to use a
2015 or later Thinkpad X1 Carbon with 2560x1440 screen running
.Ox
and chromium at 150% scaling with the dwm window manager.
This will enable the main menu to line up just right.
.Sh SEE ALSO
.Xr intro 1 ,
.Xr honk 8
.Sh STANDARDS
.Pp
.Lk https://www.w3.org/TR/activitypub/ "ActivityPub"
.Pp
.Lk https://www.w3.org/TR/activitystreams-vocabulary/ "Activity Vocabulary"
.Sh HISTORY
Started March 2019.
.Sh AUTHORS
.An Ted Unangst Lk https://honk.tedunangst.com/u/tedu @tedu@honk.tedunangst.com
.Sh CAVEATS
Completing some operations, such as subscribing to new honkers, requires an
aptitude for clipboard use and tab switching along with a steady hand.
For the most part, these are infrequent operations, but they are also the
first operations new users encounter.
This is not ideal.
.Pp
The ActivityPub standard is subject to interpretation, and not all
implementations are as enlightened as
.Nm .
.Sh BUGS
It's a feature.


docs/changelog.txt
changelog

=== next

--- Add Reactions.

+++ Rename react to badonk

+ Quick fix to hide all images.

+ Allow resending follow requests.

+ Improved search query parsing.

+ Tables

+ Reduce retries talking to dumb servers.

=== 0.8.6 Sartorial Headpiece

++ Import command now supports the elephant in the room.

+ Minimal support for Move activity.

+ deluser command.

+ Configurable avatar colors.

+ Optional pleroma color scheme for the home sick...

+ Rebalance colors slightly. Looks a little fresher now?

+ Add unplug command for servers that have dropped off the net.

+ Add notes field to honkers to document their downfall.

+ Add notes field to filters for record keeping.

+ Negated search -terms.

+ A raw sendactivity API action for the bold.

+ More flexible meme names.

=== 0.8.5 Turnkey Blaster

+ Codenames in changelog.

+ Fix some bugs that may have interfered with federation.

+ Add some re: re: re: to replies.

+ Set an avatar. If you must.

+ Try a little harder to recover from httpsig failures.

+ Add cite tag for block quote attributions.

+ @media print styles.

+ Disable overscroll (pull down) refresh.

+ Can never seem to version the changelog correctly.

=== 0.8.4

+ Fix bug preventing import of keys

+ Option to switch map links to Apple.

=== 0.8.3

- mistag.

=== 0.8.2 Game Warden

++ Import command to preserve those embarssassing old posts from Twitter.

++ Add a limited /api for the robotrons.

+ Resource usage stats on about page.

+ Unveil and pledge restrictions on OpenBSD.

+ Lists supported in markdown.

+ Rewrite admin console to avoid large dependencies.

+ "Bug" fixes.

=== 0.8.1

++ Make it easier to upgrade by decoupling data dir from ".".

+ Timestamps displayed in server time with TZ.

+ version command to print current version.

+ Amend changelog for 0.8.0 to include omitted elements:
	Syntax highlighting for code blocks.
	Something resembling an actual manual.

=== 0.8.0 Ordinary Octology

+++ Add Honk Filtering and Censorship System (HFCS).

+++ Editing honks (Update activity).

++ Subscribe to hashtags.

++ Search. I hate it already.

++ Hashtags that work?

++ Dynamic refresh and page switching without reloads.

++ Reply control. Ack replies to show them on the site.

+ Allow PDF attachments. For serious business only.

+ "untag me" button to mute part of a thread.

+ Inline images in posts. Send and receive.

+ Somewhat functional admin console (TTY).

+ More JS free fallbacks for some basic functions.

+ Add chpass command.

+ Improved honker management.

+ Better markdown output.

+ Times for events.

+ Split media database into separate blob.db.

+ Location checkin. Welcome to the... danger zone!

+ Quick mention @alias.

+ Image descriptions.

+ Unbonking.

+ More robust retries for fetching objects.

+ Don't decode excessively large images and run out of memory.

+ Syntax highlighting for code blocks.

+ Something resembling an actual manual.

- Sometimes the cached state of the @me feed becomes unsynced.
	Acked status may display incorrectly.

=== 0.7.7 More 7 Than Ever

+ Add another retry to workaround pixelfed's general unreliability.

+ Attached images are not lost when previewing.

- Remove sensitivity to spicy peppers.

+ Keep reply to setting during preview.

+ Increase max thread retrieval depth to 10.

=== 0.7.6

+ Fix a bug where upgrades would not complete in one step.

=== 0.7.5

+ Fix a bug (introdcued 0.7.4) preventing new user creation from working.

+ Semi flexible URL patterns to allow transition from other software.

+ Improved ActivityPub parsing conformance for better compat with others.

+ Add server name to user agent.

+ What may be considered UI improvements.

=== 0.7.4

+ Ever more bug fixes.

+ Collapse posts based on custom regex match.

+ Tonks are now honk backs.

+ Show both avatars for bonks. Other minor refinements to UI.

+ Minimal support for Video activity and PeerTube compat.

+ Support for some user selectable styling. Currently, skinny column mode.

+ webp image transcoding.

=== 0.7.3

+ Better fedicompat so bonks are visible to pleroma followers.

=== 0.7.2

+ Add the funzone. Minor other UI tweaks.

=== 0.7.1

+ Fix bug preventing unfollow from working.

=== 0.7.0 Father Mother Maiden Crone Honker Bonker Zonker

+++ Auto fetching and inlining of hoots.

++ A new xzone to view and import data not otherwise visible.

++ Preview before honking.

++ Some extra commands for better database retention management.

++ A changelog.

+ Default robots.txt.

+ Misc UI touchups.

+ Read only support for qonks.

+ About page.

+ More reliable (retries) meta messages such as follow requests.

+ Better thread support for missing context.

+ Upgrade image library for cleaner screenshots.

+ Not all summaries need labels.

+ Add max-width for video tag.

=== 0.6.0 Sixy Delights

Most records from this time of primitive development have been lost.

=== 0.5.0 Halfway to Heaven

=== 0.4.0 Fore Score

=== 0.3.0 Valorous Varaha


honk.go
//
// Copyright (c) 2019 Ted Unangst <tedu@tedunangst.com>
//
// Permission to use, copy, modify, and distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

package main

import (
	"flag"
	"fmt"
	"html/template"
	"log"
	notrand "math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"humungus.tedunangst.com/r/webs/httpsig"
)

var softwareVersion = "develop"

func init() {
	notrand.Seed(time.Now().Unix())
}

type WhatAbout struct {
	ID      int64
	Name    string
	Display string
	About   string
	Key     string
	URL     string
	Options UserOptions
	SecKey  httpsig.PrivateKey
}

type UserOptions struct {
	SkinnyCSS  bool   `json:",omitempty"`
	OmitImages bool   `json:",omitempty"`
	Avatar     string `json:",omitempty"`
	MapLink    string `json:",omitempty"`
	Reaction   string `json:",omitempty"`
}

type KeyInfo struct {
	keyname string
	seckey  httpsig.PrivateKey
}

const serverUID int64 = -2

type Honk struct {
	ID       int64
	UserID   int64
	Username string
	What     string
	Honker   string
	Handle   string
	Oonker   string
	Oondle   string
	XID      string
	RID      string
	Date     time.Time
	URL      string
	Noise    string
	Precis   string
	Format   string
	Convoy   string
	Audience []string
	Public   bool
	Whofore  int64
	Replies  []*Honk
	Flags    int64
	HTPrecis template.HTML
	HTML     template.HTML
	Style    string
	Open     string
	Donks    []*Donk
	Onts     []string
	Place    *Place
	Time     *Time
	Mentions []Mention
}

type Mention struct {
	Who   string
	Where string
}

type OldRevision struct {
	Precis string
	Noise  string
}

const (
	flagIsAcked    = 1
	flagIsBonked   = 2
	flagIsSaved    = 4
	flagIsUntagged = 8
	flagIsReacted  = 16
)

func (honk *Honk) IsAcked() bool {
	return honk.Flags&flagIsAcked != 0
}

func (honk *Honk) IsBonked() bool {
	return honk.Flags&flagIsBonked != 0
}

func (honk *Honk) IsSaved() bool {
	return honk.Flags&flagIsSaved != 0
}

func (honk *Honk) IsUntagged() bool {
	return honk.Flags&flagIsUntagged != 0
}

func (honk *Honk) IsReacted() bool {
	return honk.Flags&flagIsReacted != 0
}

type Donk struct {
	FileID int64
	XID    string
	Name   string
	Desc   string
	URL    string
	Media  string
	Local  bool
}

type Place struct {
	Name      string
	Latitude  float64
	Longitude float64
	Url       string
}

type Duration int64

func (d Duration) String() string {
	s := time.Duration(d).String()
	if strings.HasSuffix(s, "m0s") {
		s = s[:len(s)-2]
	}
	if strings.HasSuffix(s, "h0m") {
		s = s[:len(s)-2]
	}
	return s
}

func parseDuration(s string) time.Duration {
	didx := strings.IndexByte(s, 'd')
	if didx != -1 {
		days, _ := strconv.ParseInt(s[:didx], 10, 0)
		dur, _ := time.ParseDuration(s[didx:])
		return dur + 24*time.Hour*time.Duration(days)
	}
	dur, _ := time.ParseDuration(s)
	return dur
}

type Time struct {
	StartTime time.Time
	EndTime   time.Time
	Duration  Duration
}

type Honker struct {
	ID     int64
	UserID int64
	Name   string
	XID    string
	Handle string
	Flavor string
	Combos []string
	Meta   HonkerMeta
}

type HonkerMeta struct {
	Notes string
}

type SomeThing struct {
	What  int
	XID   string
	Owner string
	Name  string
}

const (
	SomeNothing int = iota
	SomeActor
	SomeCollection
)

var serverName string
var dataDir = "."
var viewDir = "."
var iconName = "icon.png"
var serverMsg template.HTML
var aboutMsg template.HTML
var loginMsg template.HTML

func ElaborateUnitTests() {
}

func unplugserver(hostname string) {
	db := opendatabase()
	xid := fmt.Sprintf("%%https://%s/%%", hostname)
	db.Exec("delete from honkers where xid like ? and flavor = 'dub'", xid)
	db.Exec("delete from doovers where rcpt like ?", xid)
}

func main() {
	flag.StringVar(&dataDir, "datadir", dataDir, "data directory")
	flag.StringVar(&viewDir, "viewdir", viewDir, "view directory")
	flag.Parse()
	args := flag.Args()
	cmd := "run"
	if len(args) > 0 {
		cmd = args[0]
	}
	switch cmd {
	case "init":
		initdb()
	case "upgrade":
		upgradedb()
	case "version":
		fmt.Println(softwareVersion)
		os.Exit(0)
	}
	db := opendatabase()
	dbversion := 0
	getconfig("dbversion", &dbversion)
	if dbversion != myVersion {
		log.Fatal("incorrect database version. run upgrade.")
	}
	getconfig("servermsg", &serverMsg)
	getconfig("aboutmsg", &aboutMsg)
	getconfig("loginmsg", &loginMsg)
	getconfig("servername", &serverName)
	getconfig("usersep", &userSep)
	getconfig("honksep", &honkSep)
	prepareStatements(db)
	switch cmd {
	case "admin":
		adminscreen()
	case "import":
		if len(args) != 4 {
			log.Fatal("import username mastodon|twitter srcdir")
		}
		importMain(args[1], args[2], args[3])
	case "debug":
		if len(args) != 2 {
			log.Fatal("need an argument: debug (on|off)")
		}
		switch args[1] {
		case "on":
			setconfig("debug", 1)
		case "off":
			setconfig("debug", 0)
		default:
			log.Fatal("argument must be on or off")
		}
	case "adduser":
		adduser()
	case "deluser":
		if len(args) < 2 {
			fmt.Printf("usage: honk deluser username\n")
			return
		}
		deluser(args[1])
	case "chpass":
		chpass()
	case "cleanup":
		arg := "30"
		if len(args) > 1 {
			arg = args[1]
		}
		cleanupdb(arg)
	case "unplug":
		if len(args) < 2 {
			fmt.Printf("usage: honk unplug servername\n")
			return
		}
		name := args[1]
		unplugserver(name)
	case "ping":
		if len(args) < 3 {
			fmt.Printf("usage: honk ping from to\n")
			return
		}
		name := args[1]
		targ := args[2]
		user, err := butwhatabout(name)
		if err != nil {
			log.Printf("unknown user")
			return
		}
		ping(user, targ)
	case "run":
		serve()
	case "backend":
		backendServer()
	case "test":
		ElaborateUnitTests()
	default:
		log.Fatal("unknown command")
	}
}
